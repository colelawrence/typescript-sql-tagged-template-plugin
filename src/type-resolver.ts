import * as ts from "typescript/lib/tsserverlibrary";

const getSymbolTableValues = (table: ts.SymbolTable) => {
	const values = [];
	const iter = table.values();
	for (let next = iter.next(); !next.done; next = iter.next()) {
		values.push(next.value);
	}

	return values;
};

const getTypeName = (
	typescript: typeof ts,
	checker: ts.TypeChecker,
	type: ts.Type
) =>
	checker.typeToString(
		type,
		undefined,
		typescript.TypeFormatFlags.InTypeAlias |
			typescript.TypeFormatFlags.NoTruncation
	);

const getTypeArgumentNames = (
	typescript: typeof ts,
	checker: ts.TypeChecker,
	type: ts.Type
) => {
	if (type.flags & typescript.TypeFlags.Object) {
		const otype = <ts.ObjectType>type;
		if (otype.objectFlags & typescript.ObjectFlags.Reference) {
			const rtype = <ts.TypeReference>type;
			if (rtype.typeArguments) {
				return rtype.typeArguments.map(arg =>
					getTypeName(typescript, checker, arg)
				);
			}
		}
	}

	return [];
};

const getNodeTypeName = (
	typescript: typeof ts,
	checker: ts.TypeChecker,
	node: ts.Node
): string => {
	const type = checker.getTypeAtLocation(node);
	if (type.symbol && type.symbol.flags & typescript.SymbolFlags.Interface) {
		const typeArgumentNames = getTypeArgumentNames(typescript, checker, type);
		if (type.symbol.escapedName === "Array") {
			return `Array<${typeArgumentNames.join(", ")}>`;
		} else {
			let name = "{ ";
			if (type.symbol.members) {
				const members = getSymbolTableValues(type.symbol.members);
				const typeParamTypes: { [key: string]: string | undefined } = members
					.filter(member => member.flags & typescript.SymbolFlags.TypeParameter)
					.map((member, i) => [
						member.escapedName as string,
						typeArgumentNames[i]
					])
					.reduce((acc, [name, type]) => ({ ...acc, [name]: type }), {});

				for (const member of members.filter(
					member => member.flags & typescript.SymbolFlags.Property
				)) {
					let memberType = getNodeTypeName(
						typescript,
						checker,
						member.valueDeclaration
					);
					if (typeParamTypes[memberType]) {
						memberType = typeParamTypes[memberType]!;
					}

					name += `${member.escapedName}: ${memberType}; `;
				}
			}
			name += "}";
			return name;
		}
	} else {
		return getTypeName(typescript, checker, type);
	}
};

export class TypeResolver {
	constructor(
		private readonly typescript: typeof ts,
		private readonly getTypeChecker: () => ts.TypeChecker
	) {}

	getType(node: ts.Node): string {
		const checker = this.getTypeChecker();
		return getNodeTypeName(this.typescript, checker, node);
	}
}
