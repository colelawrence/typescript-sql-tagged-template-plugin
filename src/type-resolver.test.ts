import { readdirSync } from "fs";
import { join as joinPath } from "path";
import * as ts from "typescript/lib/tsserverlibrary";
import { TypeResolver } from "./type-resolver";

const findTemplateExpression = (source: ts.SourceFile) => {
	const visit = (node: ts.Node): ts.Expression | undefined =>
		ts.isTemplateExpression(node)
			? node.templateSpans[0].expression
			: ts.forEachChild(node, visit);
	return ts.forEachChild(source, visit);
};

describe(TypeResolver, () => {
	const testProjectDir = joinPath(__dirname, "__testproject__");
	const files = readdirSync(testProjectDir);
	for (const file of files) {
		test(file, () => {
			const path = joinPath(testProjectDir, file);
			const program = ts.createProgram({
				rootNames: [path],
				options: {
					strict: true
				}
			});
			const source = program.getSourceFile(path);
			if (!source) {
				throw Error("Source file not found");
			}

			const node = findTemplateExpression(source);
			if (!node) {
				throw Error("Template expression not found");
			}

			const checker = new TypeResolver(ts, () => program.getTypeChecker());
			const type = checker.getType(node);
			expect(type).toMatchSnapshot();
		});
	}
});
