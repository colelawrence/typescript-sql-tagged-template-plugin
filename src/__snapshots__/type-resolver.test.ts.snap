// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeResolver array.ts 1`] = `"Array<\\"todo\\" | \\"in_progress\\" | \\"finished\\">"`;

exports[`TypeResolver generic-interface.ts 1`] = `"{ prop1: string; prop2: number; prop3: \\"todo\\" | \\"in_progress\\" | \\"finished\\"; }"`;

exports[`TypeResolver interface.ts 1`] = `"{ prop1: string; prop2: number; prop3: \\"todo\\" | \\"in_progress\\" | \\"finished\\"; }"`;

exports[`TypeResolver simple.ts 1`] = `"string"`;

exports[`TypeResolver union.ts 1`] = `"\\"todo\\" | \\"in_progress\\" | \\"finished\\""`;
